### DISCUSSION TITLE
Only pin Composer dependencies if already pinned; bump otherwise

### DISCUSSION THREAD

**Discussion by colinodell** [renovatebot/renovate#14155] [Suggest an Idea]
I'd like to propose altering the behavior of Composer's `auto` range strategy to better align with how PHP maintainers use Composer in practice.

## Proposal

Revise the `auto` range strategy for Composer to only `pin` versions if they're already pinned; otherwise, if a range is currently specified, the `bump` strategy would be used instead.

## Context

Renovate automatically uses the `pin` strategy for certain packages you usually want to keep up-to-date (all dev dependencies, and non-dev dependencies for applications), and the logic behind that is sound. However, it's not uncommon for PHP repositories to have a mix of pinned and caret-based constraints for those packages, and for a good reason:

 - The pinned versions signal that devs really want to stay on the current version regardless of new updates; any changes should be considered carefully.
 - Caret-based constraints (like `^1.2.3`) signal that the newer versions should be safe

Think about it from the perspective of what would happen if a user ran `composer update` manually on such a repository:

 - Pinned versions would not be updated
 - Caret-based constraints would install the newer version

I think that Renovate should respect the user's preference and better align the behavior of how `composer update` works when following the `auto` strategy.

So instead of having the `auto` strategy always `pin` these, I propose that it **only** `pin` already-pinned versions and instead `bump` anything defined as a range.

## Alternative Approaches

I work at an organization with hundreds of repositories and would prefer a solution that implements this approach as a reusable preset that all repos can extend.

Although it is possible to manually achieve the desired behavior on a per-repository basis, there's no way to apply a single custom preset across multiple repositories that varies this option based on whether a project is an "application" or a library, since that data does not seem to be exposed as a targetable option.  If this can be exposed somehow that would provide a viable alternative to this problem.

## Breaking Changes

This proposal will **not** cause any breaking changes for existing users, since the `auto` strategy will have already pinned their dependencies and thus those will continue to have newer versions pinned.

Only new repositories (which have never been updated by Renovate before) would see a difference in behavior.

## Implementation

See #14151 for the proposed implementation

**Comment by rarkins** [comment]
I disagree pretty much 100% with what you propose.

1. There is no such concept as pinning an already pinned version. To pin means to convert from a range to a fixed version. Any thing else is called *updating*
2. Keeping pinned versions pinned and unpinned versions unpinned is what rangeStrategy=replace already does. Use that instead of `auto` if that's what you want
3. The logic of mimicking `composer update` is mistaken. If you spend your life running `composer update` then you'll rapidly be stuck behind latest versions, which is the opposite of Renovate's purpose

**Reply by colinodell** [reply]
> 1. There is no such concept as pinning an already pinned version. To pin means to convert from a range to a fixed version. Any thing else is called updating

Thanks for that clarification!

> 2. Keeping pinned versions pinned and unpinned versions unpinned is what rangeStrategy=replace already does. Use that instead of auto if that's what you want

Unfortunately, that's not what I want.

**My ultimate goal is to use `bump` in virtually all cases** unless the dependency is a direct (`require`) dependency of a repository that is a library where `widen` or `replace` is more appropriate.

I could achieve this within a single application repository with a configuration like this:

```json
{
    "rangeStrategy": "bump"
}
```

Or within a library like this:

```json5
{
  "packageRules": [
    {
      "matchDepTypes": ["dependencies"],
      "rangeStrategy": "replace" // or "widen"
    }
  ],
  "packageRules": [
    {
      "matchDepTypes": ["devDependencies"],
      "rangeStrategy": "bump"
    }
  ]
}
```

But I can't match based on the [`composerJsonType`](https://github.com/renovatebot/renovate/blob/917f30e/lib/manager/composer/range.ts#L31-L39), so I can't safely define a single shared preset for both applications and libraries to use. However, the `auto` strategy already knows about the `composerJsonType`, which is why I presented that specific implementation.  Perhaps that was an XY problem.

As I noted in the original post above, defining two separate presets could technically work, but my repositories are already configured to use a single, centralized preset so having to make hundreds of PRs to change that would be a hassle I'm hoping to avoid.

Since you're not open to my original suggestion, would you be open to adding a new `matchRepositoryType` option that could differentiate between an application and a library?  That would also solve my issue.

> 3. The logic of mimicking composer update is mistaken. If you spend your life running composer update then you'll rapidly be stuck behind latest versions, which is the opposite of Renovate's purpose

I wasn't proposing to directly mimic `composer update` or to run it manually. Rather, it was to make Renovate's `auto` strategy retain the semantic meaning behind the original constraint ranges when performing automated updates as one possible solution that would achieve the behavior I'm looking for.

**Reply by rarkins** [reply]
@colinodell thanks very much for the extended details. Just wanted to let you know that I have marked this for follow-up but will need more time, likely next week

**Reply by rarkins** [reply]
If we can increase the complexity of `auto` and change some results in a way which *everyone* would agree with, then I think we could consider this. But if this is what you'd call an *opinionated* approach that a significant percent of users would not agree with, then that's a harder ask.

Another possibility would be if we standardize the field `packageFileType` and allow you to match that in package rules? Then you have control over it, and don't need to worry about `auto`?

**Reply by colinodell** [reply]
> But if this is what you'd call an opinionated approach that a significant percent of users would not agree with, then that's a harder ask.

Yeah, that is a tricky one to judge...

My suggestion is based on how I've seen most PHP developers update constraints when **not** using automated dependency management tools.  I have never seen a PHP developer "pin" a dependency in their `composer.json` unless they explicitly wanted to stay on an old version indefinitely due to BC breaks.  Most developers will either "bump", "widen", or leave their constraints as-is in 99% of cases.  That's why I personally found that pinning behavior to be surprising and unwanted.

If `auto` were to be changed, I think the least controversial change would be to not "pin" any dependency that's not already pinned and to use a sensible strategy similar to their current constraints.  However...

> Another possibility would be if we standardize the field packageFileType and allow you to match that in package rules? Then you have control over it, and don't need to worry about auto?

I really like that idea! It would take the opinion out of Renovate and empower users to implement exactly what they want.  It would completely satisfy my need without needing to change `auto`.

**Reply by colinodell** [reply]
It looks like there's an existing issue related to exposing the package type: https://github.com/renovatebot/renovate/issues/12839
